src/bodywork/workflow_execution.py

```diff
@@ -490,16 +506,63 @@ def parse_dockerhub_image_string(image_string: str) -> Tuple[str, str]:
     return image_name, image_tag
 
 
-def _print_logs_to_stdout(namespace: str, job_or_deployment_name: str) -> None:
+def _compute_optimal_job_timeout(batch_stages: List[BatchStageConfig]) -> int:
+    """Compute the optimal timeout for job monitoring.
+
+    :param namesapce: The target namespace for the job.
+    :param batch_stages: The desired configuration for the jobs.
+    :param returns: The optimal timeout (in seconds).
+    """
+    job_timeouts = [
+        max(1, stage.retries) * stage.max_completion_time for stage in batch_stages
+    ]
+    return int(max(job_timeouts) + TIMEOUT_GRACE_SECONDS)
+
+
+def _compute_optimal_deployment_timeout(
+    namespace: str, service_stages: List[ServiceStageConfig]
+) -> int:
+    """Compute the optimal timeout for deployment monitoring.
+
+    If a deployment is rolling-out for the first time, then wait for
+    twice the max configured startup time. If a deployment is updating,
+    then compute how long it will take to replace all pods using the
+    rolling update strategy.
+
+    The the max configured startup time has been floored at 60s, because
+    installing just Flask alone takes this long and it would be easy to
+    incorrectly estimate this.
+
+    :param namesapce: The target namespace for the deployments.
+    :param service_stages: The desired configuration for the incoming
+        deployments.
+    :param returns: The optimal timeout (in seconds).
+    """
+    new_pod_rate = K8S_MAX_SURGE + K8S_MAX_UNAVAILABLE
+    deployment_timeouts = [
+        (
+            stage.max_startup_time
+            if not k8s.is_existing_deployment(namespace, stage.name)
+            else ceil(stage.replicas / new_pod_rate) * max(60, stage.max_startup_time)
+        )
+        for stage in service_stages
+    ]
+    return int(2 * max(deployment_timeouts) + TIMEOUT_GRACE_SECONDS)
```

Comment:
Might be a good idea to add a one liner in the docs somewhere to explain about the min startup time etc.

Reply:
Agreed. I've made a note of this for when it's released.